### PR TITLE
lldb 16.0.6

### PR DESCRIPTION
**Pull request by xhochy**
Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by xhochy** [comment]
@isuruf @h-vetinari This fails in Linux and locally on `osx-arm64` (my laptop, no cross-compile) with the following:

```
+ lldb --version
lldb: symbol lookup error: /home/conda/feedstock_root/build_artifacts/lldb_1695731998689/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/bin/../lib/liblldb.so.16: undefined symbol: _Py_Dealloc
Tests failed for lldb-16.0.6-py310h444f4ca_0.conda - moving package to /home/conda/feedstock_root/build_artifacts/broken
```

This is related to the `libpython` linkage patch applied here. I don't know how to solve this as
1. We would like to avoid the linkage in the `lldb` Python module.
2. We would like to have the linkage in the case of the `lldb` binary.

For most of the other issues in this PR, I have an idea but this one isn't easy.

**Comment by xhochy** [comment]
@conda-forge-admin please rerender

**Comment by h-vetinari** [comment]
Basically I picked this up because I had not been aware of this feedstock, saw that it was severely outdated, and since I'm taking care of much of the LLVM space, I thought I'd tackle it.

IMO it's worth noting that this feedstock didn't support osx-arm64 so far, so that's a somewhat separate issue from updating the version to revive the feedstock. That said, obviously it's worth fixing if we can!

> This is related to the `libpython` linkage patch applied here. I don't know how to solve this [...]

I looked at the patch, and I'm surprised about "but [the lldb] python executable is linked to static python"; since we're not depending on `libpython-static` in this recipe, how could anything link statically?

So I checked the timelines, and the [introduction](https://github.com/conda-forge/lldb-feedstock/pull/16) of the patch here predates the [split](https://github.com/conda-forge/python-feedstock/commit/9a4f37217c34c10899826292d565c5e2ca80247c) of python into shared and static. However, the fact that this feedstock got built successfully until 3.10 means that everything must have somehow worked when things got linked only dynamically.

Not sure what happened in the meantime (though to be fair, we're also absorbing about 3 years worth of upstream changes). Perhaps it's worth trying to drop that patch?

**Comment by xhochy** [comment]
If I remove the patch, I get a segmentation fault:

```
import: 'lldb'
Process 55589 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x10)
    frame #0: 0x0000000101e4d3a0 libpython3.10.dylib`new_dict + 36
libpython3.10.dylib`new_dict:
->  0x101e4d3a0 <+36>: ldr    x8, [x8, #0x10]
    0x101e4d3a4 <+40>: ldr    w9, [x8, #0x2c40]
    0x101e4d3a8 <+44>: cbz    w9, 0x101e4d430           ; <+180>
    0x101e4d3ac <+48>: sub    w9, w9, #0x1
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x10)
  * frame #0: 0x0000000101e4d3a0 libpython3.10.dylib`new_dict + 36
    frame #1: 0x0000000102b8d250 liblldb.16.0.6.dylib`PyInit__lldb + 108
    frame #2: 0x0000000100174e18 python`_imp_create_dynamic + 1188
    frame #3: 0x00000001000abb44 python`cfunction_vectorcall_FASTCALL + 200
    frame #4: 0x0000000100144740 python`_PyEval_EvalFrameDefault + 26824
```

I can try and enforce linkage to `libpython3.10.a`.

**Comment by xhochy** [comment]
Static linkage leads to the same traceback:

```
Process 59815 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x10)
    frame #0: 0x0000000102df60b4 liblldb.16.0.6.dylib`new_dict + 36
liblldb.16.0.6.dylib`new_dict:
->  0x102df60b4 <+36>: ldr    x8, [x8, #0x10]
    0x102df60b8 <+40>: ldr    w9, [x8, #0x2c40]
    0x102df60bc <+44>: cbz    w9, 0x102df6148           ; <+184>
    0x102df60c0 <+48>: sub    w9, w9, #0x1
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x10)
  * frame #0: 0x0000000102df60b4 liblldb.16.0.6.dylib`new_dict + 36
    frame #1: 0x00000001025e6264 liblldb.16.0.6.dylib`PyInit__lldb + 108
```

**Comment by h-vetinari** [comment]
Something seems to be going wrong with how the linked lib name/path is constructed on windows:
```
[769/972] cmd.exe /C "cd . && %BUILD_PREFIX%\Library\bin\cmake.exe -E vs_link_exe --intdir=tools\argdumper\CMakeFiles\lldb-argdumper.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\mt.exe --manifests  -- C:\PROGRA~1\MICROS~2\2022\ENTERP~1\VC\Tools\MSVC\1429~1.301\bin\HostX64\x64\link.exe /nologo tools\argdumper\CMakeFiles\lldb-argdumper.dir\argdumper.cpp.obj  /out:bin\lldb-argdumper.exe /implib:lib\lldb-argdumper.lib /pdb:bin\lldb-argdumper.pdb /version:0.0 /machine:x64 /STACK:10000000 /INCREMENTAL:NO /subsystem:console  %PREFIX%\Library\lib\LLVMSupport.lib  LLVMD:/bld/lldb_1695839122061/_h_env/libs/python310.lib.lib  psapi.lib  shell32.lib  ole32.lib  uuid.lib  advapi32.lib  %PREFIX%\Library\lib\z.lib  %PREFIX%\Library\lib\zstd.lib  delayimp.lib  -delayload:shell32.dll  -delayload:ole32.dll  %PREFIX%\Library\lib\LLVMDemangle.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
FAILED: bin/lldb-argdumper.exe 
cmd.exe /C "cd . && %BUILD_PREFIX%\Library\bin\cmake.exe -E vs_link_exe --intdir=tools\argdumper\CMakeFiles\lldb-argdumper.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\mt.exe --manifests  -- C:\PROGRA~1\MICROS~2\2022\ENTERP~1\VC\Tools\MSVC\1429~1.301\bin\HostX64\x64\link.exe /nologo tools\argdumper\CMakeFiles\lldb-argdumper.dir\argdumper.cpp.obj  /out:bin\lldb-argdumper.exe /implib:lib\lldb-argdumper.lib /pdb:bin\lldb-argdumper.pdb /version:0.0 /machine:x64 /STACK:10000000 /INCREMENTAL:NO /subsystem:console  %PREFIX%\Library\lib\LLVMSupport.lib  LLVMD:/bld/lldb_1695839122061/_h_env/libs/python310.lib.lib  psapi.lib  shell32.lib  ole32.lib  uuid.lib  advapi32.lib  %PREFIX%\Library\lib\z.lib  %PREFIX%\Library\lib\zstd.lib  delayimp.lib  -delayload:shell32.dll  -delayload:ole32.dll  %PREFIX%\Library\lib\LLVMDemangle.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
LINK: command "C:\PROGRA~1\MICROS~2\2022\ENTERP~1\VC\Tools\MSVC\1429~1.301\bin\HostX64\x64\link.exe /nologo tools\argdumper\CMakeFiles\lldb-argdumper.dir\argdumper.cpp.obj /out:bin\lldb-argdumper.exe /implib:lib\lldb-argdumper.lib /pdb:bin\lldb-argdumper.pdb /version:0.0 /machine:x64 /STACK:10000000 /INCREMENTAL:NO /subsystem:console %PREFIX%\Library\lib\LLVMSupport.lib LLVMD:/bld/lldb_1695839122061/_h_env/libs/python310.lib.lib psapi.lib shell32.lib ole32.lib uuid.lib advapi32.lib %PREFIX%\Library\lib\z.lib %PREFIX%\Library\lib\zstd.lib delayimp.lib -delayload:shell32.dll -delayload:ole32.dll %PREFIX%\Library\lib\LLVMDemangle.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST:EMBED,ID=1" failed (exit code 1104) with the following output:
LINK : fatal error LNK1104: cannot open file 'LLVM%PREFIX%\libs\python310.lib.lib'
```
I mean `LLVM%PREFIX%` and `.lib.lib`

**Comment by xhochy** [comment]
@isuruf @h-vetinari This is ready for review.

File: recipe/bld.bat [hunk 1]
```diff
@@ -15,19 +19,19 @@ cmake -G "Ninja" ^
     -DLLVM_TARGETS_TO_BUILD=X86 ^
     -DLLVM_TEMPORARILY_ALLOW_OLD_TOOLCHAIN=ON ^
     -DLLDB_ENABLE_PYTHON=ON ^
-    -DLLDB_PYTHON_RELATIVE_PATH=../Lib/site-packages ^
-    -DPYTHON_HOME=%PREFIX% ^
+    "-DLLDB_PYTHON_RELATIVE_PATH:PATH=..\Lib\site-packages" ^
```
Comment (h-vetinari) [inline recipe/bld.bat:22]: Style nit
```suggestion
    -DLLDB_PYTHON_RELATIVE_PATH:PATH="..\Lib\site-packages" ^
```

File: recipe/build.sh [hunk 2]
```diff
@@ -11,10 +14,23 @@ else
...
+fi
+
+if [[ "${target_platform}" != "${build_platform}" ]]; then
+  NATIVE_FLAGS="-DCMAKE_C_COMPILER=$CC_FOR_BUILD;-DCMAKE_CXX_COMPILER=$CXX_FOR_BUILD;-DCMAKE_C_FLAGS=-O2;-DCMAKE_CXX_FLAGS=-O2"
+  NATIVE_FLAGS="${NATIVE_FLAGS};-DCMAKE_EXE_LINKER_FLAGS=;-DCMAKE_MODULE_LINKER_FLAGS=;-DCMAKE_SHARED_LINKER_FLAGS="
+  NATIVE_FLAGS="${NATIVE_FLAGS};-DCMAKE_STATIC_LINKER_FLAGS=;-DCMAKE_PREFIX_PATH=$BUILD_PREFIX"
+  CMAKE_ARGS="${CMAKE_ARGS} -DCROSS_TOOLCHAIN_FLAGS_NATIVE=${NATIVE_FLAGS}"
+
+  SP_PATH=${SP_DIR//$PREFIX}
+  SP_PATH=${SP_PATH:1}
```
Comment (h-vetinari) [inline recipe/build.sh:28]: This could use a comment perhaps?

**Review by h-vetinari — COMMENTED** [review]
Thanks for pushing this all the way to green! :)
I don't have much to offer in terms of review, but from what I can tell, the changes look good.

I'm wondering if we should retry 15.0.7 now that we know better what to fix, in order to have a version that matches our current compiler stack? Though we could also do this on a branch later, if it even becomes necessary (we're also probably moving to LLVM 16 soonish).

Otherwise I'd be slightly tempted to streamline the commit history a bit, but that's just me. :)

File: recipe/bld.bat (hunk 1 above)
Comment (xhochy) [reply recipe/bld.bat:22]: Let's hope this doesn't make a difference (it is actually something differently on Windows)

**Comment by xhochy** [comment]
Happy to backport to 15.x but we're anyways moving to the 16.x clang stack at the moment so I think we're fine.

**Pull request merged**